src/Network/HTTP/Affjax.purs

```diff
@@ -117,6 +123,41 @@ delete u = affjax $ defaultRequest { method = DELETE, url = u }
 delete_ :: forall e. URL -> Affjax e Unit
 delete_ = delete
 
+-- | Retry a request with exponential backoff, timing out optionally after a specified number of milliseconds.
+retry :: forall e a b. (Requestable a) => Maybe Int -> (AffjaxRequest a -> Affjax (avar :: AVAR | e) b) -> (AffjaxRequest a -> Affjax (avar :: AVAR | e) b)
+retry milliseconds run req = do
+  failureVar <- makeVar
+  let loop = go failureVar
+  case milliseconds of
+    Nothing -> loop 1
+    Just milliseconds -> do
+      respVar <- makeVar
+      loopHandle <- forkAff $ loop 1 >>= putVar respVar <<< Just
+      timeoutHandle <-
+        forkAff <<< later' milliseconds $ do
+          putVar respVar Nothing
+          loopHandle `cancel` error "Cancel"
+      result <- takeVar respVar
+      case result of
+        Nothing ->
+          takeVar failureVar
+        Just resp -> pure resp
+  where
+    assert200 resp =
+      case resp.status of
+        StatusCode 200 -> Right resp
+        _ -> Left resp
+
+    go failureVar n = do
+      result <- run req
+      case assert200 result of
```

Comment:
:+1:


-- 
John A. De Goes
Follow me @jdegoes

On August 3, 2015 at 5:33:55 PM, Jonathan Sterling (notifications@github.com) wrote:

In src/Network/HTTP/Affjax.purs:

> +          putVar respVar Nothing
> +          loopHandle `cancel` error "Cancel"
> +      result <- takeVar respVar
> +      case result of
> +        Nothing ->
> +          takeVar failureVar
> +        Just resp -> pure resp
> +  where
> +    assert200 resp =
> +      case resp.status of
> +        StatusCode 200 -> Right resp
> +        _ -> Left resp
> +
> +    go failureVar n = do
> +      result <- run req
> +      case assert200 result of
Sounds like the proper thing to do here, then, is to use attempt and then retry accordingly? And then, if we are never able to contact the server within the timeout, we should bubble up the error.

—
Reply to this email directly or view it on GitHub.